### PR TITLE
Enhancement: Redaction of sensitive spans

### PR DESCRIPTION
**Pull request by Harnoor-se7en**
- Removes useless grafana metric for the count of redacted attributes.
- Add generic functionality to redact using regex and key match. 

New Redaction config is the list of below config:
```
{
        name = <Name of PII/PCI field>: String
        regexString = <regexString>: String
        keys = <List of Strings>: List
        type = <Should be either PII or PCI>: String
}
```

- The tag value will be replaced by the value of the config key 'name' with the prefix 'redacted'.
- A new Tag key will be added - 'redacted.pii.keys' or 'redacted.pii.keys' where pii/pci depends on the config key 'type'. The value of this new tag key will be the count of type keys redacted.
- If regexString is empty, the Tag value will be redacted if the tag key is present in the 'List of keys' of the config.
- If the List of keys is empty, then the Tag value will be redacted by regexMatch against the regex provided in the config.

Example 1:

```
{
        name = "accountNo"
        regexString = "[0-9]{9,18}",
        keys = ["PayeeAccountNo", "payeeaccountno", "PayerAccountNo", "payee_account"]
        type = "PII"
}
```

The above config means span redaction will happen when all the below conditions are met: 
1.  Regex match is true against tag value.
2. and the matched tag value's key is present in the list of keys provided in config.

Example 2:
```
{
        name = "phoneNo"
        regexString = "^(?:(?:\\+|0{0,2})91(\\s*[\\-]\\s*)?|[0]?)?[789]\\d{9}$",
        keys = []
        type = "PCI"
}
```
The above config means span redaction will happen when all the below conditions are met: 
1.  Regex match is true against tag value.

Example 3:
   ```
 {
        name = "http"
        regexString = ""
        keys = ["http.url", "http.method", "authorization"]
        type = "PCI"
}

```
The above config means span redaction will happen when all the below conditions are met: 
1.  Tag key of the span is present in the above tag keys.

### CODE REVIEW

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 1]
```diff
@@ -90,25 +83,16 @@ public static JaegerSpanNormalizer get(Config config) {
...
-          redactedAttributeValue =
-              AttributeValue.newBuilder()
-                  .setValue(SpanNormalizerConstants.PII_FIELD_REDACTED_VAL)
-                  .build();
-        }
+      if (config.hasPath(SpanNormalizerConstants.SPAN_REDACTION_CONFIG_KEY)) {
+        config.getConfigList(SpanNormalizerConstants.PII_PCI_CONFIG_KEY).stream()
+            .map(
+                conf ->
+                    new PIIPCIField(
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:89]: Either use chaining or builder to create new objects.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 2]
```diff
@@ -160,9 +140,8 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
           .normalize(jaegerSpan, tenantIdHandler.getTenantIdProvider().getTenantIdTagKey())
           .ifPresent(rawSpanBuilder::setResource);
 
-      // redact PII tags, tag comparisons are case insensitive (Resource tags are skipped)
-      if (redactedAttributeValue != null) {
-        sanitiseSpan(rawSpanBuilder);
+      if (!PIIPCIFields.isEmpty()) {
+        redactSpan(rawSpanBuilder);
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: Rename to `redactAttributes`.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 3]
```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
 
-  private void sanitiseSpan(Builder rawSpanBuilder) {
-
+  // redact PII tags, tag comparisons are case-insensitive (Resource tags are skipped)
+  private void redactSpan(Builder rawSpanBuilder) {
     try {
       var attributeMap = rawSpanBuilder.getEvent().getAttributes().getAttributeMap();
-      var spanServiceName = rawSpanBuilder.getEvent().getServiceName();
+
+      Integer piiFieldsCount = 0;
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: `int`

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.Optional;
+import java.util.Set;
+import org.apache.commons.lang3.StringUtils;
+import org.hypertrace.core.datamodel.AttributeValue;
+
+public class PIIPCIField {
+
+  private String name;
+  private Set<String> tagKeySet;
+  private RegexInfo regexInfo;
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java:61]: Better move `RegexInfo` as a public static class inside this class. This keeps related things close-by.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 5]
```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
+        if (piiPciField.getRegexInfo().isPresent()) {
+          matcher = piiPciField.getRegexInfo().get().getRegexPattern().matcher("");
+        }
+        PIIPCIFieldType piiPciFieldType = piiPciField.getPiiPciFieldType();
+        for (String tagKey : tagKeys) {
+          /* A Simple regex like PAN no. can have false positives.
+             Hence, we only redact tag val when tagKey is present in possible list of regexTagKeySet.
+             If regexTagKeySet is empty, we redact the tag val always.
+          */
+          if (!piiPciField.getTagKeySet().isEmpty()
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: Move this predicate to an appositely named method.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 6]
```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
-      var spanServiceName = rawSpanBuilder.getEvent().getServiceName();
+
+      Integer piiFieldsCount = 0;
+      Integer pciFieldsCount = 0;
+
       Set<String> tagKeys = attributeMap.keySet();
+      for (PIIPCIField piiPciField : PIIPCIFields) {
+        Matcher matcher = null;
+        if (piiPciField.getRegexInfo().isPresent()) {
+          matcher = piiPciField.getRegexInfo().get().getRegexPattern().matcher("");
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:168]: Why not create the `matcher` after line 183?

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 7]
```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
-                  logSpanRedaction(tagKey, spanServiceName, PIIMatchType.KEY);
-                  attributeMap.put(tagKey, redactedAttributeValue);
-                });
-      } catch (Exception e) {
-        LOG.error("An exception occurred while sanitising spans with key match: ", e);
-      }
+          boolean containsSensitiveData = false;
+          if (piiPciField.getRegexInfo().isPresent()) {
+            assert matcher != null;
+            matcher.reset(attributeMap.get(tagKey).getValue());
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:185]: Why do you need a reset when you're creating a new matcher object for each tag key-val basically?

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 6 above)
Comment (suddendust) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:168]: Ignore this comment.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 7 above)
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:183]: This'll have thread safety issues if `sanitiseSpan` is called on different threads.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 8]
```diff
@@ -175,103 +154,69 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
+            if (matcher.find()) {
+              containsSensitiveData = true;
+            }
+          } else {
+            // this condition implies that match type of PII/PCI field is KEY based.
+            containsSensitiveData = true;
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:189]: Please rename this field.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 7 above)
Comment (suddendust) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:183]: No not really, I see a new matcher is created every time. So for an average of 300k spans per second and 10 regex rules, we'll be creating 3M matcher objects per second, is it?

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 8 above)
Comment (suddendust) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:189]: Also, what if the `tagKeySet` of the PII/PCI field is empty?

**Review by suddendust — CHANGES_REQUESTED** [review]
Review, address as discussed.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 3 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: why?

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 7 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:183]: Yes

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 8 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:189]: This is fixed.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 1 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:89]: I don't think it is required over here. What's wrong with this approach though? 
I see this pattern already being used by the traceable team: [https://github.com/razorpay/hypertrace-ingester/blob/7b03729ab4930cabc7525ae97a7312907d9acafc/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/SpanFilter.java#L80](https://github.com/razorpay/hypertrace-ingester/blob/7b03729ab4930cabc7525ae97a7312907d9acafc/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/SpanFilter.java#L80)

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 2 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: ok

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 3 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: aah okay I just read this: [https://stackoverflow.com/questions/5199359/why-do-people-still-use-primitive-types-in-java](https://stackoverflow.com/questions/5199359/why-do-people-still-use-primitive-types-in-java)

Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: done

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java (hunk 4 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java:61]: Done. Have moved the enum inside this class too.

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 5 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java]: ok

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java (hunk 6 above)
Comment (Harnoor-se7en) [reply span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:168]: ok

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      }
+      this.replacementValue =
+          AttributeValue.newBuilder().setValue(REDACTED_FIELD_PREFIX + name).build();
+      this.piiPciFieldType = PIIPCIFieldType.valueOf(type);
+    }
+  }
+
+  private boolean isValidConfig(String regex, List<String> tagKeyList, String type) {
+    if (StringUtils.isEmpty(regex) && tagKeyList.isEmpty()) {
+      throw new RuntimeException("Both Regex and TagKeyList cannot be empty.");
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/redaction/PIIPCIField.java:80]: Better: Use `IllegalArgumentException`

File: span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java [hunk 10]
```diff
@@ -175,106 +153,75 @@ private Callable<RawSpan> getRawSpanNormalizerCallable(
...
-                      k ->
-                          registerCounter(
-                              SPAN_REDACTED_ATTRIBUTES_COUNTER,
-                              Map.of("matchType", PIIMatchType.REGEX.toString())))
-                  .increment();
-              logSpanRedaction(tagKey, spanServiceName, PIIMatchType.REGEX);
-              attributeMap.put(tagKey, redactedAttributeValue);
+          if (containsSensitiveData) {
+            if (piiPciFieldType == PIIPCIField.PIIPCIFieldType.PII) {
+              piiFieldsCount += 1;
```
Comment (suddendust) [inline span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanNormalizer.java:194]: Why not just do `++`?

**Review by suddendust — APPROVED** [review]
LGTM, please do a quick perf run before deployment.

**Pull request merged**